lib/class-wp-theme-json.php

```diff
@@ -0,0 +1,953 @@
+							),
+							'fontSize'                     => true,
+							'lineHeight'                   => true,
+						),
+					)
+				),
+			)
+		);
+		foreach ( $blocks as $block_name => $block_type ) {
+			if (
+				! property_exists( $block_type, 'supports' ) ||
+				empty( $block_type->supports ) ||
+				! is_array( $block_type->supports ) ) {
+
+				// Skips blocks that don't declare support.
+				//
+				// TODO: what if there are blocks that don't support
+				// any style but still need the settings passed down?
+				continue;
+			}
+
+			$block_supports = array();
+			foreach ( self::PROPERTIES_METADATA as $key => $metadata ) {
+				if ( $this->get_from_path( $block_type->supports, $metadata['block_json'] ) ) {
+					$block_supports[] = $key;
+				}
+			}
+
+			/*
+			 * Assign the selector for the block.
+			 *
+			 * Some blocks can declare multiple selectors:
+			 *
+			 * - core/heading represents the H1-H6 HTML elements
+			 * - core/list represents the UL and OL HTML elements
+			 * - core/group is meant to represent DIV and other HTML elements
+			 *
+			 * Some other blocks don't provide a selector,
+			 * so we generate a class for them based on their name:
+			 *
+			 * - 'core/group' => '.wp-block-group'
+			 * - 'my-custom-library/block-name' => '.wp-block-my-custom-library-block-name'
+			 *
+			 * Note that, for core blocks, we don't add the `core/` prefix to its class name.
+			 * This is for historical reasons, as they come with a class without that infix.
+			 *
+			 */
+			if (
+				isset( $block_type->supports['__experimentalSelector'] ) &&
+				is_string( $block_type->supports['__experimentalSelector'] )
+			) {
+				self::$blocks_metadata[ $block_name ] = array(
+					'selector'  => $block_type->supports['__experimentalSelector'],
+					'supports'  => $block_supports,
+					'blockName' => $block_name,
+				);
+			} elseif (
+				isset( $block_type->supports['__experimentalSelector'] ) &&
+				is_array( $block_type->supports['__experimentalSelector'] )
+			) {
+				foreach ( $block_type->supports['__experimentalSelector'] as $key => $selector ) {
+					self::$blocks_metadata[ $key ] = array(
+						'selector'  => $selector,
+						'supports'  => $block_supports,
+						'blockName' => $block_name,
+					);
+				}
+			} else {
+				self::$blocks_metadata[ $block_name ] = array(
+					'selector'  => '.wp-block-' . str_replace( '/', '-', str_replace( 'core/', '', $block_name ) ),
+					'supports'  => $block_supports,
+					'blockName' => $block_name,
+				);
+			}
+		}
+
+		return self::$blocks_metadata;
+	}
+
+	/**
+	 * Normalize the subtree according to the given schema.
+	 * This function modifies the given input by removing
+	 * the nodes that aren't valid per the schema.
+	 *
+	 * @param string $key Key of the subtree to normalize.
+	 * @param array  $input Whole tree to normalize.
+	 * @param array  $schema Schema to use for normalization.
+	 */
+	private function process_key( $key, &$input, $schema ) {
```

Comment:
Why do we need to remove the nodes that aren't valid per the schema? When outputting CSS could we simply ignore these nodes?

Reply:
Right, we can do validation at entering-time or at outputting-time. I chose to do it on entering because that simplifies all the internal code for processing the tree (stylesheet processing but also settings, etc). It's also safer, as we are strict on which things we allow and which things we don't, which is important in case attackers are able to hook into the theme.json in the server.